fabric-provider-api-v1/src/main/java/net/fabricmc/fabric/api/provider/v1/ApiProvider.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.provider.v1;
+
+/**
+ * Supplies API instances associated with a game object.
+ *
+ * <p>Consumers can obtain an instance of the provider
+ * without knowing how or where it is implemented by using
+ * {@link ApiProviderAccess#getAccess()}.
+ *
+ * <p>This decoupling of provider access from provider
+ * implementation is the primary purpose of this module.
+ *
+ * <p>Implementations are expected to extend this interface to
+ * add methods with additional parameters as needed.
+ *
+ * <p>Allocation and management of any state, along with thread
+ * safety, are fully delegated to implementations.
+ *
+ * @param <P> Identifies the API provider type
+ * @param <A> Identifies the API type
+ */
+@FunctionalInterface
```

Comment:
It is an intentional suggestion that sub-types remain lambda-friendly, with the existing method and any overloads getting default implementations. If a provider function is going to be very complex it will likely need a context object and in that case a functional interface should also be viable.

That said, implementations aren't forced to maintain the functional interface contract if it doesn't work for them.

Another option I had considered was to remove ApiProvider entirely and allow any type at all.  The system seemed more difficult to understand that way but it would be even more flexible - would make it easier to have providers that a "single hop" - the mapping function returns the API instance directly if the minimal built-in access patterns are sufficient.

Reply:
Seems reasonable to keep the annotation then.